charabia/src/segmenter/utils.rs

```diff
@@ -23,15 +29,40 @@ impl<'fst> FstSegmenter<'fst> {
                 return None;
             }
 
-            let length = match find_longest_prefix(self.words_fst, to_segment.as_bytes()) {
+            let mut length = match find_longest_prefix(self.words_fst, to_segment.as_bytes()) {
                 Some((_, length)) => length,
                 None => {
-                    // if no sequence matches, we return the next character as a lemma.
-                    let first = to_segment.chars().next().unwrap();
-                    first.len_utf8()
+                    if self.allow_char_split {
+                        // if no sequence matches, we return the next character as a lemma.
+                        to_segment.chars().next().unwrap().len_utf8()
+                    } else {
+                        // if splitting is not allowed, return the whole input
+                        let result = to_segment;
+                        to_segment = "";
+                        return Some(result);
+                    }
                 }
             };
 
+            if let Some(min_len) = self.min_length {
+                // enforce minimum lemma length if specified
+                if length < min_len && to_segment.len() > length {
+                    length = min_len.min(to_segment.len());
+                }
+
+                // prevent left over lemmas with a length fewer than min_len
+                if to_segment.len() - length < min_len {
+                    length = to_segment.len();
+                }
+            }
+
+            // ensure the length is a valid character boundary
+            length = to_segment
+                .char_indices()
+                .nth(length)
+                .map(|(idx, _)| idx)
+                .unwrap_or(to_segment.len());
```

Comment:
Because the length is a byte length, we can't use nth because it count characters and not bytes:
```suggestion
            // ensure the length is a valid character boundary
            length = to_segment
                .char_indices()
                .find(|(idx, _)| idx >= length)
                .map(|(idx, _)| idx)
                .unwrap_or(to_segment.len());
```